### PR TITLE
Help user on Redshift debugging, remove red herring error message on pure ASCII files

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This adds some help to allow a user to more directly query `stl_load_errors` upon COPY errors.

It also eliminates a false error message when someone tries to load a plain ASCII file.

### CODE REVIEW

File: records_mover/db/redshift/loader.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from ...records.processing_instructions import ProcessingInstructions
 import sqlalchemy
 from sqlalchemy.schema import Table
+import psycopg2
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:8]: F401 'psycopg2' imported but unused

File: records_mover/db/redshift/loader.py [hunk 2]
```diff
@@ -79,7 +80,61 @@ def load(self,
                                empty_as_null=True,
                                **redshift_options)  # type: ignore
             logger.info(f"Starting Redshift COPY from {directory}...")
-            self.db.execute(copy)
+# analytics=> \d+ stl_load_errors
+#                                            Table "pg_catalog.stl_load_errors"
+#      Column      |            Type             | Collation | Nullable | Default | Storage  | Stats target | Description
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:85]: E501 line too long (121 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 3]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+# -----------------+-----------------------------+-----------+----------+---------+----------+--------------+-------------
+#  userid          | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:87]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 4]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  slice           | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:88]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 5]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  tbl             | integer                     |           | not null |         | plain    |              |
+#  starttime       | timestamp without time zone |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:90]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 6]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  session         | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:91]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 7]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  query           | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:92]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 8]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  filename        | character(256)              |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:93]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 9]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  line_number     | bigint                      |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:94]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 10]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  colname         | character(127)              |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:95]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 11]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  type            | character(10)               |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:96]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 12]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  col_length      | character(10)               |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:97]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 13]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  position        | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:98]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 14]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  raw_line        | character(1024)             |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:99]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 15]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  raw_field_value | character(1024)             |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:100]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 16]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  err_code        | integer                     |           | not null |         | plain    |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:101]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 17]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+#  err_reason      | character(100)              |           | not null |         | extended |              |
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:102]: E501 line too long (109 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 18]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+# Replica Identity: ???
+
+# analytics=>
+
+# analytics=> select * from stl_load_errors where session=11400;
+# -[ RECORD 1 ]---+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:108]: E501 line too long (1044 > 100 characters)

File: records_mover/db/redshift/loader.py [hunk 19]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+# position        | 45
+# raw_line        | WI,WI-8848269,20200317.0,,,Absentee,20200416,20200407.0,,,
+# raw_field_value | 20200407.0
+# err_code        | 1207
+# err_reason      | Invalid digit, Value '.', Pos 8, Type: Long
+
+# analytics=> \q
+
+
+            # TODO: sqlalchemy.exc.InternalError: (psycopg2.errors.InternalError_) Load into table 'ts_early_and_absentee_0416' failed.  Check 'stl_load_errors' system table for details.
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:129]: E303 too many blank lines (2)

Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:129]: E501 line too long (186 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:129]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/db/redshift/loader.py [hunk 20]
```diff
@@ -79,7 +80,61 @@ def load(self,
...
+            redshift_pid = self.db.execute("SELECT pg_backend_pid();").fetchone()[0]
+            try:
+                self.db.execute(copy) # TODO: handle and report back
```
Comment (bluelabsbutler) [inline records_mover/db/redshift/loader.py:132]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: tests/unit/db/redshift/test_loader.py [hunk 21]
```diff
@@ -105,3 +106,47 @@ def test_load_non_s3(self,
                                empty_as_null=True,
                                abc=123)
         self.mock_db.execute.assert_called_with(mock_copy)
+
+
+    @patch('records_mover.db.redshift.loader.CopyCommand')
```
Comment (bluelabsbutler) [inline tests/unit/db/redshift/test_loader.py:111]: E303 too many blank lines (2)

File: tests/unit/db/redshift/test_loader.py [hunk 22]
```diff
@@ -105,3 +106,47 @@ def test_load_non_s3(self,
...
+                             mock_CopyCommand):
+        mock_schema = Mock(name='schema')
+        mock_table = Mock(name='table')
+        mock_load_plan = Mock(name='load_plan')
+        mock_load_plan.records_format = Mock(name='records_format', spec=DelimitedRecordsFormat)
+        mock_load_plan.records_format.hints = {}
+        mock_directory = Mock(name='directory')
+        mock_directory.scheme = 'mumble'
+
+        mock_temp_s3_loc = self.mock_temporary_s3_directory_loc.return_value.__enter__.return_value
```
Comment (bluelabsbutler) [inline tests/unit/db/redshift/test_loader.py:128]: F841 local variable 'mock_temp_s3_loc' is assigned to but never used

File: tests/unit/db/redshift/test_loader.py [hunk 23]
```diff
@@ -105,3 +106,47 @@ def test_load_non_s3(self,
...
+        mock_s3_directory = mock_directory.copy_to.return_value
+        mock_s3_directory.scheme = 's3'
+
+        mock_aws_creds = mock_s3_directory.loc.aws_creds.return_value
```
Comment (bluelabsbutler) [inline tests/unit/db/redshift/test_loader.py:132]: F841 local variable 'mock_aws_creds' is assigned to but never used

File: records_mover/db/redshift/loader.py [hunk 24]
```diff
@@ -79,7 +79,19 @@ def load(self,
...
+            redshift_pid: int = self.db.execute("SELECT pg_backend_pid();").scalar()
+            try:
+                self.db.execute(copy)
+            except sqlalchemy.exc.InternalError:
+                # Upon a load erorr, we receive:
+                #
+                #  sqlalchemy.exc.InternalError:
+                #  (psycopg2.errors.InternalError_) Load into table 'tablename'
+                #    failed.  Check 'stl_load_errors' system table for details.
+                logger.warning("Caught load error - "
```
Comment (cwegrzyn) [inline records_mover/db/redshift/loader.py:91]: ❤️

File: records_mover/records/hints.py [hunk 25]
```diff
@@ -52,6 +52,8 @@ def cant_handle_hint(fail_if_cant_handle_hint: bool, hint_name: str, hints: Reco
     # But let's be ready if they change their minds:
     'UTF-8': 'UTF8',
     'Windows-1252': 'CP1252',
+    # even if the only data it saw was in ASCII, let's be ready to see more
+    'ascii': 'UTF8',
```
Comment (cwegrzyn) [inline records_mover/records/hints.py:56]: can't recall-- any case things to worry about here?

Comment (cwegrzyn) [reply records_mover/records/hints.py:56]: ("case" as in upper/lower)

File: tests/unit/db/redshift/test_loader.py [hunk 26]
```diff
@@ -105,3 +106,44 @@ def test_load_non_s3(self,
...
+
+        mock_s3_directory = mock_directory.copy_to.return_value
+        mock_s3_directory.scheme = 's3'
+
+        mock_redshift_options = {'abc': 123}
+        mock_redshift_copy_options.return_value = mock_redshift_options
+        mock_copy = mock_CopyCommand.return_value
+        mock_s3_directory.loc.url = 's3://foo/bar/baz/'
+
+        def db_execute(command):
```
Comment (cwegrzyn) [inline tests/unit/db/redshift/test_loader.py:135]: Nice use of mock side effects!

File: records_mover/records/hints.py (hunk 25 above)
Comment (vinceatbluelabs) [reply records_mover/records/hints.py:56]: Nah, this is the correct case on both sides - it made the error message go away (chardet returns "ascii"), and Python spells UTF-8 as "UTF-8".

File: records_mover/db/redshift/loader.py (hunk 24 above)
Comment (vinceatbluelabs) [reply records_mover/db/redshift/loader.py:91]: Yeah, I've been wanting this forever - happy to have figured out the pattern to make it happen!

**Pull request merged**